### ISSUE TITLE
Longevity of IPFS content referenced in issues ( Github, Mailing lists, etc )

### ISSUE THREAD

**Issue by mib-kd743naq** [ipfs/community#327]
**NOTE**: This issue is **orthogonal / the opposite of** *github issues on IPFS*

A non-trivial portion of bugreports / issues / design documents currently available on github etc **do refer to IPFS file content** ( `Qm...` ). A staggering portion of **this content does not currently resolve**: either being lost for good or sitting on nodes not currently online. To make matters worse much of this content is a direct product of ProtocolLabs members, which has implications on the perceived health of the product.

Some sort of quick-and-dirty solution needs to be implemented as soon as feasible to systematically crawl and pin content across at least github issues/pulls, due to the high visibility of resolution failures.
There would be implications with potential for abuse of the pinning service / moderation evasion, but these should be dealt with as they arise, and not get in the way of "the simplest thing that can/should exist today"

/cc @VictorBjelkholm @Mr0grog @whyrusleeping

**Comment by lgierth** [comment]
Yeah that's a thing I've been thinking about too. Instead of a crawler, there should be a webhook listener that scans for certain events and pins the respective stuff. In this case here mostly for CIDs in comment texts, but similar tools could add new commits on our repos to IPFS.

**Comment by lgierth** [comment]
Commits that updates `.gx/lastpubver` and `package.json` in go repos should also result in pins.

**Comment by mib-kd743naq** [comment]
And another tangent to this issue: the [blogpost by @diasdavid](http://blog.daviddias.me/2015/12/08/stellar-module-management) referenced in the [ipfs-npm docs](https://www.npmjs.com/package/ipfs-npm#resources) is no longer available :(

**Comment by VictorBjelkholm** [comment]
Yeah, I think there are two areas where we need to immediately improve. First is something that goes through *all* (yeah...) existing content within our organizations and checks if it's actually pinned somewhere. We should be able to run this on some interval and it should report issues with links to the affected content.

Second is moving forward, a webhook-service that analyses all new content and makes sure it's pinned.

Side-note: @mib-kd743naq you can now find that blogpost here: http://daviddias.me/blog/stellar-module-management/ seems @diasdavid changed the structure of the blog.

**Comment by Mr0grog** [comment]
@lgierth @VictorBjelkholm Sounds like we are all agreed on needing to do something for this. Where does this work need to happen and how can we best track it? Who can take responsibility for it?

>> Instead of a crawler, there should be a webhook listener that scans for certain events and pins the respective stuff.
> 
> I think there are two areas… First is something that goes through all (yeah...) existing content within our organizations and checks if it's actually pinned somewhere… second is moving forward, a webhook-service that analyses all new content

Ha, seems like we have different views here. I think it might make sense to focus first on what we can do going forward — *new* posts/content should usually not have the problem we’re trying to prevent (content not being resolvable), so they will be pretty easy to handle. Old stuff will be much more likely to be unresolvable, and every unresolvable link will be a complicated project to try and find. We may not even always be able to solve it. That makes new stuff a much simpler problem.

**Comment by Mr0grog** [comment]
@lgierth @VictorBjelkholm is there any part of this a community member can write or prototype that you can then fit into PL’s infrastructure? e.g:

- Library of functions that, given a GitHub API object for posts/commits/etc., identify a list of IPFS content paths/references?
- Webhook handling server that pipes things through the above functions and then…
    - Calls some pluggable function/module?
    - Calls another webhook address that can be configured?
    - Sends commands to pinbot? (any reference for how to do that?)
- Script that, given a repo, pulls all the various objects (e.g. issue comments, commits) (maybe which types of objects is configurable) that might have IPFS content paths and runs them through the above functions and then does similar to the webhook server described above?

If someone could work on those, is there a preferred language or toolset that makes it easy for you to integrate?

Trying to think of how folks in the community can pitch in on this, since I know you time is limited and I’m sure there are people who’d love to help.

**Comment by VictorBjelkholm** [comment]
I feel like the most helpful entrypoint would be to just have a github-adapter to work with https://github.com/ipfs/ipfs-hubot (which should basically be plug-and-play, except a regex to grab hashes from posts/commits/etc). ipfs-hubot is also deployed already (somewhere, unsure of where actually) so would just have to add the adapter and deploy the changes.
